misk-zookeeper/src/main/kotlin/misk/clustering/zookeeper/ZkLease.kt

```diff
@@ -0,0 +1,200 @@
+ * check is made against the current process hash to determine if the lease should still hold the
+ * lease, and if not, it will release the lease, allowing the correctly hashed process to acquire
+ * the lease subsequently.
+ *
+ * <p> Once acquired, the lease will generally remain held as long as the lease holder process
+ * remains running. However, under certain conditions, such as a network partition or a
+ * substantial full JVM pause, the lease may expire if zookeeper doesn't hear from client before
+ * the zookeeper session timeout passes. This means that under certain circumstances, the
+ * leadership durability is tied to the zookeeper session timeout.
+ *
+ * <p>
+ * Users of this lease framework should avoid creating tasks that have a duration longer than
+ * the session timeout or they risk having a split brain running two leader tasks on different hosts
+ * simultaneously.
+ */
+internal class ZkLease(
+  ownerName: String,
+  private val manager: ZkLeaseManager,
+  private val leaseResourceName: String,
+  override val name: String
+) : Lease {
+
+  enum class Status {
+    UNKNOWN,
+    NOT_HELD,
+    HELD,
+    CLOSED
+  }
+
+  @GuardedBy("lock") private var status: Status = Status.UNKNOWN
+  private val lock = ReentrantLock()
+  private val leaseData = ownerName.toByteArray(Charsets.UTF_8)
+  private val leaseZkPath = name.asZkPath
+
+  override fun checkHeld(): Boolean {
+    try {
+      if (!manager.isRunning || !manager.isConnected) {
+        return false
+      }
+
+      return lock.withLock {
+        if (shouldHoldLease()) {
+          when (status) {
+            // We already own the lease
+            Status.HELD -> true
+
+            // We don't own the lease but should - try to acquire it
+            Status.UNKNOWN, Status.NOT_HELD -> tryAcquireLeaseNode()
+
+            // The entire leasing system is shutting down so we shouldn't attempt to hold leases
+            Status.CLOSED -> false
+          }
+        } else {
+          // We shouldn't own the lease
+          status = if (release(true)) Status.NOT_HELD else Status.UNKNOWN
+          false
+        }
+      }
+    } catch (e: Exception) {
+      log.error(e) { "unexpected exception checking if lease $name is held" }
+      return false
+    } catch (e: Error) {
+      log.error(e) { "unexpected error checking if lease $name is held" }
+      throw e
+    }
+  }
+
+  fun close() {
+    lock.withLock {
+      if (status == Status.HELD) release(true)
+      status = Status.CLOSED
+    }
+  }
+
+  fun connectionLost() {
+    lock.withLock {
+      // We're disconnected from the cluster, so we don't know the state of the lease. Zk will
+      // automatically expire the lease if we remain disconnected for too long, and in the
+      // meanwhile we should act as though the lease has been lost
+      if (status != Status.CLOSED) status = Status.UNKNOWN
+    }
+  }
+
+  private fun release(guaranteeDelete: Boolean = true): Boolean {
```

Comment:
is the guaranteeDelete flag ever used as false?

Reply:
oh eh only in a previous iteration. removing